odoo/addons/base/models/ir_cron.py

```diff
@@ -439,17 +506,24 @@ def _run_job(cls, job):
                     if status == CompletionStatus.FULLY_DONE and progress.deactivate:
                         job['active'] = False
                 finally:
+                    done, remaining = progress.done, progress.remaining
                     progress.timed_out_counter = 0
+                    progress.duration = (time.time() - start_time)
                     timed_out_counter = 0
-                    job_cr.commit()
-                _logger.info('Job %r (%s) processed %s records, %s records remaining',
-                             job['cron_name'], job['id'], progress.done, progress.remaining)
+                    job_cr.commit()  # ensure we have no more leftovers
+                    _logger.info('Job %r (%s) processed %s records, %s records remaining',
+                        job['cron_name'], job['id'], done, remaining)
                 if status in (CompletionStatus.FULLY_DONE, CompletionStatus.FAILED):
                     break
+                if end_time <= time.time():
```

Comment:
Subtract the time of the last run to avoid going over the soft limit?
@Julien00859 We may still go over the limit because it is just an estimation, we are just going to be closer.